Initial Loki implementation of cloudsc2-nl on CPU

piotrows opened the pull request:
Current code version verifies - in addition to nl/tl/ad -  correctly on a single ATOS CPU for Loki idem and scc and scc-hoist variants. The following major code modifications present here are:
- Copy of the legacy cloudsc-loki variant in development of ml
- Inline of Cuadj* and turn cloudsc2/satur into modules, following cloudsc-loki example
- Synchronization of the CMake building strategy for Loki variant with the clouadsc-loki reference
- Global parameters are being passed via argument line rather than USE, following cloudsc reference and previous PR

mlange05 committed:
COMMON: Rename library to `cloudsc2-common-lib`

mlange05 committed:
CLOUDSC2: Add interface header imports for cloudsc2

mlange05 committed:
Loki: Initial CLOUDSC2-NL Loki setup with "idem" mode

mlange05 committed:
CLOUDSC2: Add OpenACC detection to cmake configurations

mlange05 committed:
Loki: Add Loki-SCC configuration and necessary fixes for CLOUDSC2-NL

piotrows committed:
Initial commit, files from experimental loki branch by ml copied

piotrows committed:
CMake setup synchronized with cloudsc-loki, satur etc. files need to be sorted out yet

piotrows committed:
cmake fixed, stuck on timer.mod

piotrows committed:
merge

piotrows committed:
Merge commit '2e6edbf0fc72dfdf9ea5fc1753783f342e69d165' into napz-loki-nl

piotrows committed:
timer.mod problem

piotrows committed:
Bring back env.sh

piotrows committed:
fix to common lib name, fix missing inlined CUADJ temporaries declarations

piotrows committed:
Removal of unused files, add derived types to argument list and synchronize INTERFACE accordingly

piotrows committed:
Fix misisng derv types in satur

piotrows committed:
Trying to overcome INTERFACE problem, mac+gcc fails

piotrows committed:
add missing header

piotrows committed:
idem nl seems to compile on ATOS

piotrows committed:
scc compiles as well, cannot run yet on ATOS

piotrows committed:
Reactivation of nl/tl/ad

piotrows committed:
Remark on Loki variant in README.md

FussyDuck commented:
[![CLA assistant check](https://bol-claassistant.ecmwf.int:443/pull/badge/signed)](https://bol-claassistant.ecmwf.int:443/ecmwf-ifs/dwarf-p-cloudsc2-tl-ad?pullRequest=8) <br/>All committers have signed the CLA.

piotrows committed:
Fix for scc-hoist, note that updated Loki is needed, Loki update in the PR by ML at the moment

mlange05 committed:
Merge branch 'develop' into napz-loki-nl

mlange05 commented on cmake/python_venv.cmake:
```diff
@@ -0,0 +1,179 @@
+##############################################################################
+#.rst:
+#
+# find_python_venv
```
Are we using this for the Loki setup, or is this just blindly copied? If not, please remove, as I think Loki does handle its own venvs.

piotrows replied:
Removed

mlange05 commented on cmake/OMP.cmake:
```diff
@@ -0,0 +1,44 @@
+if( HAVE_OMP )
+
+    try_compile(
```
This OMP detection mechanism is really just for OMP device offload, right? As nothing in this PR requires it, and since we're not testing it, I'd say we can drop this until we really need it and are ready support this properly.

mlange05 replied:
Note, the same applies to the source files for the detection test.

piotrows replied:
Removed

mlange05 commented on src/cloudsc2_nl_loki/CMakeLists.txt:
```diff
@@ -0,0 +1,197 @@
+
+function( cloudsc_xmod _TARGET )
+
+    if( TARGET clawfc AND ${LOKI_FRONTEND} STREQUAL "omni" )
```
Due to the complexity that OMNI and OFP frontends bring, I'm tempted to say we can exclude them here. We use them to test edge cases in CLOUDSC, but I think there is little gained here from carrying all of that over. Opinions? (Also paging @reuterbal ..)

reuterbal replied:
Agreed, I think we should focus on the Fparser API here and rather work on making this functional for other frontends under the hood if it is ever desired.

piotrows replied:
Omni and OFP removed

mlange05 commented on src/cloudsc2_nl_loki/CMakeLists.txt:
```diff
@@ -0,0 +1,197 @@
+
+function( cloudsc_xmod _TARGET )
+
+    if( TARGET clawfc AND ${LOKI_FRONTEND} STREQUAL "omni" )
```
Due to the complexity that OMNI and OFP frontends bring, I'm tempted to say we can exclude them here. We use them to test edge cases in CLOUDSC, but I think there is little gained here from carrying all of that over. Opinions? (Also paging @reuterbal ...)

mlange05 commented on src/cloudsc2_nl_loki/cloudsc_loki.config:
```diff
@@ -0,0 +1,51 @@
+name = 'cloudsc2'
+role = 'kernel'
+expand = true
+demote_locals = false
```
Ohhhhh? :open_mouth: Is locals demotion disabled for a reason here? Does it change/impact results?

mlange05 commented on src/cloudsc2_nl_loki/cloudsc_loki.config:
```diff
@@ -0,0 +1,51 @@
+name = 'cloudsc2'
+role = 'kernel'
+expand = true
+demote_locals = false
```
Ohhhhh? :open_mouth: Is locals demotion disabled for a reason here? Does it change/impact results?

piotrows replied:
removed demote=false, results on cpu are fine

mlange05 commented on src/cloudsc2_nl_loki/cuadjtqs.F90:
```diff
@@ -0,0 +1,281 @@
+! granted to it by virtue of its status as an intergovernmental organisation
+! nor does it submit to any jurisdiction.
+!
+SUBROUTINE CUADJTQS &
```
I believe this file has been manually inlined in the kernel, no? Shouldn't it also come out of the common lib, if this were not the case?

piotrows replied:
It is inlined, I am not sure if it should come out from the common lib, it is not the case in the reference (nl variant). Removed

mlange05 commented on src/cloudsc2_nl_loki/dwarf_cloudsc.F90:
```diff
@@ -0,0 +1,134 @@
+!ENDDO
+! Instead we do this similar computation involving values already available
+DO JK=1,GLOBAL_STATE%KLEV
+  YRECLD%CETA(JK)= GLOBAL_STATE%PAP(1,JK,1)/GLOBAL_STATE%PAPH(1,GLOBAL_STATE%KLEV+1,1)
```
[optional] As a small side not, I do wonder if this ratio wouldn't be much easier to compute on-the-fly? But this does not need to be investigated as part of this PR.

mlange05 commented on src/common/CMakeLists.txt:
```diff
@@ -1,44 +1,88 @@
-target_include_directories( cloudsc-common-lib PUBLIC $<INSTALL_INTERFACE:include> )
-target_include_directories( cloudsc-common-lib PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>)
-target_include_directories( cloudsc-common-lib PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/module>)
+target_include_directories( cloudsc2-common-lib PUBLIC $<INSTALL_INTERFACE:include> )
```
Could we go to the same generator-expression-based inline definition as we do in CLOUDSC? As in add this to the `ecbuild_add_library()` call above? 
```
    PRIVATE_INCLUDES
        $<${HAVE_HDF5}:${HDF5_Fortran_INCLUDE_DIRS}>
    PUBLIC_INCLUDES
        $<INSTALL_INTERFACE:include>
        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
        $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/module>
```

piotrows replied:
Done in the later PR, back ported.

mlange05 commented on src/common/CMakeLists.txt:
```diff
@@ -1,44 +1,88 @@
+
+
+if( HAVE_ACC )
+  # ========================================================================
```
Same comment as above; we don't need to build both lib types any more. Instead we can toggle as we do in CLOUDSC.

piotrows replied:
Done.

mlange05 reviewed (changes requested):
Ok, this looks nice on the whole, but I think a bunch of unnecessary boiler plate code has crept in here that isn't needed or can be updated now. I've not run this yet, but wanted to file first round early to get this going. I'll give this another go once some of the comments have been addressed.

piotrows committed:
Remove spurious cmake components, update common lib build code

piotrows committed:
remove inlined cuadjtqs.F90 in nl-loki, remove demote=false

piotrows committed:
Readme update

mlange05 committed:
Loki: Disable Loki processing by default

mlange05 committed:
Loki: Disbale OpenACC generation for SCC/SCC-H NL variants

mlange05 committed:
CMake: Removing legacy/compat OpenACC and OpenMP-device detection

THese are not yet supported, and should only be re-introduced once
properly tested.

mlange05 reviewed (approved):
Ok, the Loki-SCC/SCC-hoist variants are not yet GPU clean, and need proper disabling for CPU-only mode. I've added `DIRECTIVE none` to the CMake loki macros for now, which requires latest loki-main, but now builds clean without any artifacts on Intel/NVHPC for CPU.

I've also removed the CMake compat features for old-style OpenACC (this was for pre-3.16, we now require >=3.17) and the OpenMP device offload testing. This should be re-introduced once we actually test and support OpenMP GPU offload in this dwarf.
